finance-reports/mstro_reports_to_socrata.py

```diff
@@ -0,0 +1,140 @@
+import csv
+from io import StringIO
+import os
+import logging
+
+import boto3
+from sodapy import Socrata
+
+from config import EXPENSES_FIELD_MAPPING
+from config import EXPENSES_NUMERIC_COLS
+from config import MONTH_LOOKUP
+from config import REVENUE_FIELD_MAPPING
+from config import REVENUE_NUMERIC_COLS
+import utils
+
+# AWS Credentials
+AWS_ACCESS_ID = os.getenv("AWS_ACCESS_KEY")
+AWS_SECRET_ACCESS_KEY = os.getenv("AWS_SECRET_ACCESS_KEY")
+BUCKET = os.getenv("BUCKET_NAME")
+
+# Socrata Secrets
+SO_WEB = os.getenv("SO_WEB")
+SO_TOKEN = os.getenv("SO_TOKEN")
+SO_KEY = os.getenv("SO_KEY")
+SO_SECRET = os.getenv("SO_SECRET")
+EXP_DATASET = os.getenv("EXP_DATASET")
+REV_DATASET = os.getenv("REV_DATASET")
+
+
+def list_s3_files(s3_client, subdir):
+    file_names = []
+
+    # List objects in the bucket
+    response = s3_client.list_objects_v2(Bucket=BUCKET, Prefix=subdir)
+
+    # Add file names to the list
+    for obj in response.get("Contents", []):
+        file_names.append(obj["Key"])
+
+    return file_names
+
+
+def get_csv_data(s3_client, filename, field_mapping):
+    response = s3_client.get_object(Bucket=BUCKET, Key=filename)
+    csv_content = response["Body"].read().decode("utf-8")
+    csv_data = csv.DictReader(StringIO(csv_content))
+
+    # Field mapping to Socrata columns
+    mapped_data = []
+    for row in csv_data:
+        mapped_row = {}
+        for original_key, new_key in field_mapping.items():
+            mapped_row[new_key] = row.get(original_key)
+        mapped_data.append(mapped_row)
+
+    return mapped_data
+
+
+def get_fiscal_year(year, month):
+    if month >= 10:
+        fiscal_year = year + 1
+    else:
+        fiscal_year = year
+
+    return fiscal_year
+
+
+def get_fiscal_month(month):
+    if month >= 10:
+        fiscal_month = month - 9
+    else:
+        fiscal_month = month + 3
+
+    return fiscal_month
+
+
+def main():
+    s3_client = boto3.client(
+        "s3",
+        aws_access_key_id=AWS_ACCESS_ID,
+        aws_secret_access_key=AWS_SECRET_ACCESS_KEY,
+    )
+    soda = Socrata(SO_WEB, SO_TOKEN, username=SO_KEY, password=SO_SECRET, timeout=60*15,)
+
+    # Expenses
+    payload = []
+    files = list_s3_files(s3_client, "expenses/")
+    for f in files:
+        data = get_csv_data(s3_client, f, EXPENSES_FIELD_MAPPING)
+
+        for row in data:
+            row["year"] = int(f[9:13])
+            row["month"] = int(f[14:16])
+            row["department"] = int(f[-8:-4])
+
+            # Derived fields
+            row["month_name"] = MONTH_LOOKUP[row["month"]]
+            row["fiscal_year"] = get_fiscal_year(row["year"], row["month"])
+            row["fiscal_month"] = get_fiscal_month(row["month"])
+
+            # Convert empty strings to None for numeric fields
+            for key in EXPENSES_NUMERIC_COLS:
+                if row[key] == "":
+                    row[key] = None
+
+        payload.extend(data)
+
+    res = soda.replace(EXP_DATASET, payload)
```

Comment:
Every time this script runs, we have to replace the whole dataset because there are no logical IDs for each row. I have pulled data dating back to Oct 2020, so this can end up being quite large which is not ideal but I'm not sure how else to go about it.

Reply:
what makes a row unique? is it a combo of these things?

- Fund@Code
- Fund@Long Name
- Department@Dept
- Unit@Unit Code
- Object Code Category@Code
- Object Code@Code
- Object Code - Spending Plan@Code
- % Obligated

Maybe you could MD5 or something similar?

Reply:
yeah @johnclary probably year+month+FDU+object

Reply:
FWIW, the in its current state, this runs in about 8 minutes on my laptop. I'm not sure if you expect the amount of data that is processed to go up a ton in the future, so I can't say if it's worth it to use John's idea of deriving an ID from those columns. I like the idea though - very slick.

Reply:
I've followed a similar logic at the to_s3 script. Args to select what month/year (defaulting to current) and will only update the Socrata dataset with the selected month and the month previous. Also `--replace` to go through every file and upsert it.

Reply:
Also created row ID's based on the FDU and other columns